File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int len = s.length();
        int l = 0;
        Set<Character> set = new HashSet<>();
        int max = 0;
        for (int i = 0; i < len; i++) {
            int r = i;
            char c = s.charAt(r);
            while (set.contains(c)) {
                set.remove(s.charAt(l));
                l++;
            }
            set.add(c);
            max = Math.max(max, r - l + 1);
        }

        return max;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ans = 0
        l = 0
        dup = set()
        for r in range (len(s)):
            c = s[r]
            while c in dup and l <= r:
                dup.remove(s[l])
                l += 1
            ans = max(r - l + 1, ans)
            dup.add(c)
        
        return ans